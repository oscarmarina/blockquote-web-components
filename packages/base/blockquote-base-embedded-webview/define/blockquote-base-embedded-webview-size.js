import { BlockquoteBaseEmbeddedWebviewSize } from '../src/BlockquoteBaseEmbeddedWebviewSize.js';

window.customElements.define(
  'blockquote-base-embedded-webview-size',
  BlockquoteBaseEmbeddedWebviewSize,
);
