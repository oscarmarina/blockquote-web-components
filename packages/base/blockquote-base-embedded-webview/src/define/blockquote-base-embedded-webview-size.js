import {BlockquoteBaseEmbeddedWebviewSize} from '../BlockquoteBaseEmbeddedWebviewSize.js';

window.customElements.define(
  'blockquote-base-embedded-webview-size',
  BlockquoteBaseEmbeddedWebviewSize
);
