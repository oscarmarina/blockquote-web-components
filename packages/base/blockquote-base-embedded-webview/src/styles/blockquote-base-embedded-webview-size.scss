:host {
  --_host-color: var(--blockquote-base-embedded-webview-size-color, inherit);
  --_rect-height: var(--blockquote-base-embedded-webview-size-rect-height, 1.125rem);
  --_rect-size: var(--blockquote-base-embedded-webview-size-rect-size, 0.6875rem);
  --_button-border-color: var(
    --blockquote-base-embedded-webview-size-button-border-color,
    rgb(184, 184, 184)
  );
  --_button-bgcolor: var(
    --blockquote-base-embedded-webview-size-button-bgcolor,
    rgb(234, 234, 234)
  );
  --_button-bgcolor-hover: var(
    --blockquote-base-embedded-webview-size-button-bgcolor-hover,
    rgb(220, 220, 220)
  );
  --_button-bgcolor-selected-hover: var(
    --blockquote-base-embedded-webview-size-button-bgcolor-selected-hover,
    rgb(210, 210, 210)
  );

  color: var(--_host-color);
  display: block;
  box-sizing: border-box;
  pointer-events: none;
}

:host([hidden]),
[hidden] {
  display: none !important;
}

*,
*::before,
*::after {
  box-sizing: inherit;
}

button {
  appearance: none;
  background-color: transparent;
  border: none;
  color: inherit;
  cursor: pointer;
  display: inline-block;
  pointer-events: auto;
  font: inherit;
  margin: 0;
  padding: 0;
  block-size: inherit;
  text-align: inherit;
  text-decoration: none;
  text-transform: inherit;
  text-shadow: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  inline-size: auto;
}

.rect {
  overflow: hidden;
  block-size: var(--_rect-height);
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: var(--_rect-size);
  white-space: nowrap;
  text-align: center;
  letter-spacing: 0.0625rem;
  box-shadow: 0 1px 0 0 var(--_button-bgcolor);

  & [aria-hidden="true"] {
    cursor: pointer;
    letter-spacing: inherit;
    position: absolute;
  }

  & [aria-disabled="true"] {
    opacity: 0.4;

    &::after {
      content: "";
      display: block;
      block-size: 1px;
      inline-size: calc(100% + 2rem);
      position: absolute;
      inset-inline-start: 50%;
      inset-block-start: 50%;
      transform: translate(-50%, -50%);
      background-image: linear-gradient(90deg, rgb(0, 0, 0, 0), rgb(0, 0, 0, 1), rgb(0, 0, 0, 0));
    }
  }
}

button {
  position: absolute;
  background-color: var(--_button-bgcolor);
  border-inline-start: 1px solid var(--_button-border-color);
  border-inline-end: 1px solid var(--_button-border-color);

  span {
    opacity: 0;
    pointer-events: none;
    user-select: none;
  }

  &:hover {
    background-color: var(--_button-bgcolor-hover);
  }

  &[data-selected],
  &[data-selected] ~ & {
    background-color: var(--_button-bgcolor-hover);
  }

  &:hover ~ & {
    background-color: transparent;
  }

  &:not([data-selected]):hover {
    span {
      opacity: 1;
    }

    & ~ span {
      visibility: hidden;
    }
  }

  &[data-selected] ~ &:hover {
    background-color: var(--_button-bgcolor-selected-hover);
  }
}
