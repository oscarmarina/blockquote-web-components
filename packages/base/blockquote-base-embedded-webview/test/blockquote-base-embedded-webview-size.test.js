/* eslint-disable import/no-extraneous-dependencies */
import {html, fixture, assert, fixtureCleanup} from '@open-wc/testing';
import {spy} from 'sinon';
import '../define/blockquote-base-embedded-webview-size.js';

suite('BlockquoteBaseEmbeddedWebviewSize', () => {
  /**
   * @type {import('../src/index').BlockquoteBaseEmbeddedWebviewSize}
   */
  let el;
  let screenSizesButtons;
  let buttonSize3;

  const screenSizes = [
    {width: 360, height: 640, id: '360x640'},
    {width: 1024, height: 768, id: '1024x768'},
    {width: 1920, height: 1080, id: '1920x1080'},
  ];

  teardown(() => fixtureCleanup());

  suite('Default', () => {
    setup(async () => {
      el = await fixture(html`
        <blockquote-base-embedded-webview-size
          .screenSizes="${screenSizes}"></blockquote-base-embedded-webview-size>
      `);
      await el.updateComplete;

      screenSizesButtons = el.shadowRoot?.querySelectorAll('button');
      buttonSize3 = screenSizesButtons?.[2];
    });

    suite('Semantic Dom and a11y', () => {
      test('SHADOW DOM - Structure test', async () => {
        await assert.shadowDom.equalSnapshot(el);
      });

      test('LIGHT DOM - Structure test', async () => {
        await assert.lightDom.equalSnapshot(el);
      });

      test('a11y', async () => {
        await assert.isAccessible(el);
      });

      test('selectedchange event is raised', async () => {
        const spyEvent = spy();
        el.addEventListener('selectedchange', spyEvent);
        el.selected = 1;
        await el.updateComplete;
        assert.isTrue(spyEvent.called, 'selectedchange event is raised');
      });

      test('click event is raised', async () => {
        const spyEvent = spy();
        el.addEventListener('click', spyEvent);
        buttonSize3.click();
        await el.updateComplete;
        assert.isTrue(spyEvent.called, 'click event is raised');
      });

      test('screen-sizes array are sorted in descending order (by width)', () => {
        const lastIndex = Number(buttonSize3.dataset.index);
        assert.isTrue(lastIndex === screenSizesButtons.length);
      });

      test('Hidden screen size options that are too large for the container', async () => {
        assert.strictEqual(el.shadowRoot?.querySelectorAll('button[hidden]').length, 2);
        el.style.width = '1024px';
        window.dispatchEvent(new Event('resize'));
        el.requestUpdate();
        await el.updateComplete;
        assert.strictEqual(el.shadowRoot?.querySelectorAll('button[hidden]').length, 1);
      });
    });
  });

  suite('Custom selected', () => {
    setup(async () => {
      el = await fixture(html`
        <blockquote-base-embedded-webview-size
          selected="3"
          width-in-percent></blockquote-base-embedded-webview-size>
      `);
      await el.updateComplete;
    });
    suite('Semantic Dom and a11y', () => {
      test('SHADOW DOM - Structure test', async () => {
        await assert.shadowDom.equalSnapshot(el);
      });

      test('LIGHT DOM - Structure test', async () => {
        await assert.lightDom.equalSnapshot(el);
      });

      test('a11y', async () => {
        await assert.isAccessible(el);
      });
    });
  });
});
