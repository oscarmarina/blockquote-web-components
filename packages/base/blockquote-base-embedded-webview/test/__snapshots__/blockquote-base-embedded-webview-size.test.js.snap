// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BlockquoteBaseEmbeddedWebviewSize > Custom selected > LIGHT DOM - Structure test > LIGHT DOM 1`] = `
"<blockquote-base-embedded-webview-size
  selected="3"
  width-in-percent=""
>
</blockquote-base-embedded-webview-size>
"
`;

exports[`BlockquoteBaseEmbeddedWebviewSize > Custom selected > SHADOW DOM - Structure test > SHADOW DOM 1`] = `
"<div class="rect">
  <button
    data-index="1"
    hidden=""
    id="1920x1080"
    style="width: calc(100% / 1);"
  >
    <span>
      1920x1080
    </span>
  </button>
  <button
    data-index="2"
    hidden=""
    id="1536x864"
    style="width: calc(100% / 2);"
  >
    <span>
      1536x864
    </span>
  </button>
  <button
    data-index="3"
    data-selected=""
    hidden=""
    id="1366x768"
    style="width: calc(100% / 3);"
  >
    <span>
      1366x768
    </span>
  </button>
  <button
    data-index="4"
    hidden=""
    id="1280x800"
    style="width: calc(100% / 4);"
  >
    <span>
      1280x800
    </span>
  </button>
  <button
    data-index="5"
    hidden=""
    id="810x1080"
    style="width: calc(100% / 5);"
  >
    <span>
      810x1080
    </span>
  </button>
  <button
    data-index="6"
    hidden=""
    id="768x1024"
    style="width: calc(100% / 6);"
  >
    <span>
      768x1024
    </span>
  </button>
  <button
    data-index="7"
    id="414x896"
    style="width: calc(100% / 7);"
  >
    <span>
      414x896
    </span>
  </button>
  <button
    data-index="8"
    id="390x864"
    style="width: calc(100% / 8);"
  >
    <span>
      390x864
    </span>
  </button>
  <button
    data-index="9"
    id="360x800"
    style="width: calc(100% / 9);"
  >
    <span>
      360x800
    </span>
  </button>
  <span
    aria-disabled="false"
    aria-hidden="true"
  >
    1366x768
  </span>
</div>
"
`;

exports[`BlockquoteBaseEmbeddedWebviewSize > Default > LIGHT DOM - Structure test > LIGHT DOM 1`] = `
"<blockquote-base-embedded-webview-size>
</blockquote-base-embedded-webview-size>
"
`;

exports[`BlockquoteBaseEmbeddedWebviewSize > Default > SHADOW DOM - Structure test > SHADOW DOM 1`] = `
"<div class="rect">
  <button
    data-index="1"
    hidden=""
    id="1920x1080"
    style="width: 1920px;"
  >
    <span>
      1920x1080
    </span>
  </button>
  <button
    data-index="2"
    hidden=""
    id="1024x768"
    style="width: 1024px;"
  >
    <span>
      1024x768
    </span>
  </button>
  <button
    data-index="3"
    data-selected=""
    id="360x640"
    style="width: 360px;"
  >
    <span>
      360x640
    </span>
  </button>
  <span
    aria-disabled="false"
    aria-hidden="true"
  >
    360x640
  </span>
</div>
"
`;
